### PR TITLE
dolt gc --full: Implement a flag for GC which will collect everything, including the old gen.

### PR DESCRIPTION
**Pull request by reltuk**
Dolt GC is generational. By default, it moves all chunks reachable from commits into the old gen, and it leaves uncommitted data in the new gen. By default, it never revisits the chunks in the old gen. This means that if a branch with a lot of unique data exists during a GC, and then gets deleted, the storage taken up by that branch is never reclaimed by future garbage collection.

This PR implements `dolt gc --full`, which performs a collection on the entire database, ignoring the generational aspects of previous collections and visiting the entire set of live data. At the end of a successful `dolt gc --full`, only data which was reachable at the start of the run and data which has been written during the run will be present in the database.

Note: `dolt gc --full` visits every reachable chunk in the database and can be resource intensive. As a consequence of running `dolt gc --full`, any previously archived table files will be unarchived, and storage savings from the archiving will be lost. `dolt gc --full` involves copying chunks into new storage files, and fully materializing the new files before the old files are deleted; as a consequence, peak disk utilisation while  running `dolt gc --full` can be up to 2x the existing size of the database.

### CODE REVIEW

**Comment by coffeegoddd** [comment]
@reltuk __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>88a9a34</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>88a9a34</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@reltuk __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>077294f</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>077294f</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@reltuk __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>6cfac1b</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>6cfac1b</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/store/chunks/memory_store.go [hunk 1]
```diff
@@ -343,7 +343,24 @@ func (ms *MemoryStoreView) EndGC() {
 	ms.transitionToNoGC()
 }
 
-func (ms *MemoryStoreView) MarkAndSweepChunks(ctx context.Context, hashes <-chan []hash.Hash, dest ChunkStore) error {
+type msvGcFinalizer struct {
```
Comment (macneale4) [inline go/store/chunks/memory_store.go:346]: After GC, what does the memory store contain? This refactor looks like it preserves previous behavior, but I'm not sure why the previous behavior is the way it is. Educate me

File: go/store/types/value_store.go [hunk 2]
```diff
@@ -563,58 +561,107 @@ func makeBatches(hss []hash.HashSet, count int) [][]hash.Hash {
...
+		if mode == GCModeFull {
+			err = oldGenFinalizer.SwapChunksInStore(ctx)
+			if err != nil {
+				collector.EndGC()
+				return err
+			}
+		}
+
+		collector.EndGC()
+	} else if collectorOK {
```
Comment (macneale4) [inline go/store/types/value_store.go:664]: A little hard to follow the logic due to the length of the function. Possible to refactor into gcsOK func and a !gcsOK func?

File: go/store/nbs/store.go [hunk 3]
```diff
@@ -1037,6 +1037,34 @@ func (nbs *NomsBlockStore) HasMany(ctx context.Context, hashes hash.HashSet) (ha
 	return nbs.hasMany(toHasRecords(hashes))
 }
 
+func (nbs *NomsBlockStore) hasManyInSources(ctx context.Context, srcs []hash.Hash, hashes hash.HashSet) (hash.HashSet, error) {
```
Comment (macneale4) [inline go/store/nbs/store.go:1040]: Is the ctx future proofing for when we might need it? No used currently. If you think the interface needs it, I'd drop the name and make it '_' so people know that this is intentional

File: go/store/nbs/table_set.go [hunk 4]
```diff
@@ -115,6 +115,35 @@ func (ts tableSet) hasMany(addrs []hasRecord) (bool, error) {
 	return f(ts.upstream)
 }
 
+func (ts tableSet) hasManyInSources(srcs []hash.Hash, addrs []hasRecord) (remaining bool, err error) {
```
Comment (macneale4) [inline go/store/nbs/table_set.go:118]: Most functions on tableSet are documented in the chunkReader interface. This one is a one off though, so adding gDocs seems appropriate. I'm not wild about this function name being the same as the NomsBlockStore function, especially since they are different interfaces. m2c.

**Review by macneale4 — APPROVED** [review]
Nit comments and questions. No show stoppers.

File: go/libraries/doltcore/doltdb/gc_test.go [hunk 5]
```diff
@@ -139,7 +139,7 @@ func testGarbageCollection(t *testing.T, test gcTest) {
 		}
 	}
 
-	err := dEnv.DoltDB.GC(ctx, nil)
+	err := dEnv.DoltDB.GC(ctx, types.GCModeDefault, nil)
```
[the remark this answers is not in the document]
Comment (reltuk) [reply go/libraries/doltcore/doltdb/gc_test.go:142]: Yeah, I don't think these tests are worth much compared to a bats test which does the same. Obviously they're supposed to target the DoltDB.GC invocation itself, and they're much easy to attach a debugger to or whatever, but AFAICT they're not providing much in the way of protection compared to existing layers of cheese.

File: go/store/chunks/memory_store.go (hunk 1 above)
Comment (reltuk) [reply go/store/chunks/memory_store.go:346]: AFAIK, memory store is an in-memory chunk store implementation. After GC, memory store just contains all the visited chunks – that is, all the chunks whose address came through on the keepers channel.

File: go/store/nbs/store.go (hunk 3 above)
Comment (reltuk) [reply go/store/nbs/store.go:1040]: I just didn't realize it was unneeded. Removed! Thanks for that :).

File: go/store/types/value_store.go [hunk 6]
```diff
@@ -563,58 +561,107 @@ func makeBatches(hss []hash.HashSet, count int) [][]hash.Hash {
...
+
+		err := collector.BeginGC(lvs.gcAddChunk)
 		if err != nil {
 			return err
 		}
 
 		root, err := lvs.Root(ctx)
 		if err != nil {
-			newGen.EndGC()
+			collector.EndGC()
```
[the remark this answers is not in the document]
Comment (reltuk) [reply go/store/types/value_store.go:602]: It was just a lot of nesting...but in converting it I found a place where we forgot the EndGC. So...this was definitely good feedback. Thanks for it.

File: go/store/nbs/table_set.go (hunk 4 above)
Comment (reltuk) [reply go/store/nbs/table_set.go:127]: Commented. I was mirroring the |hasMany| symmetry on NomsBlockStore/tableSet, but I agree it can be a bit confusing. Not revisiting the naming for now though...

File: go/store/types/value_store.go (hunk 2 above)
Comment (reltuk) [reply go/store/types/value_store.go:664]: I agree it's a bit hard to follow, but I want to find a better way to refactor it rather than split them up, because the two branches are very intimately tied to each other and need to both be visited anytime something changes. I'll try to take another pass, but in a different PR.

**Review by macneale4 — APPROVED** [review]
```
                __/___            
          _____/______|           
  _______/_____\_______\_____     
  \              < < <       |    
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
